### ISSUE TITLE
Add utils for Interpreting F-Algebras.

### ISSUE THREAD

**Issue by MaxOw** [dhall-lang/dhall-haskell#1161]
As the title says, it would be nice to have tools for writing Interpret instance for any given recursive Haskell data type in some structured way.

What I'm proposing here is a helper function that would help Interpret any given F-Algebra `F` encoded in dhall with a following type:

```dhall
∀(a : Type) → ∀(F a → a) → a
```
Now, I have a working implementation for the required functionality in the following [gist](https://gist.github.com/MaxOw/0d0d1a42566f9dcbe3c9525250959d5c) (Although it could definitely use some love when it comes to handling alpha-renaming/name shadowing, and especially ability to use any variable name for the fix point constructor. Any hints on that front are very welcome)

So, anyway, given a recursive Haskell type of a form:

```haskell
data Expr
   = Lit Natural
   | Add Expr Expr
   | Mul Expr Expr
   deriving (Generic, Show)
```

We add a limited amount of boilerplate using `makeBaseFunctor` from `recursive-schemes` and `autoWithFix` helper function from the above mentioned gist.

```haskell
makeBaseFunctor ''Expr
deriving instance Generic (ExprF t)
instance Interpret t => Interpret (ExprF t)
instance Interpret Expr where autoWith = autoWithFix
```

And given a dhall file:

```dhall
-- F-Algebra declaration.
let ExprF
  = λ(t : Type) →
  < LitF : { _1 : Natural }
  | AddF : { _1 : t, _2 : t }
  | MulF : { _1 : t, _2 : t }
  >

-- Convenience helper.
let Expr
  = λ(t : Type)
  → λ(fix : ExprF t → t)
  → let E = ExprF t in
    { Lit = λ(x : Natural)      → fix (E.LitF { _1 = x })
    , Add = λ(a : t) → λ(b : t) → fix (E.AddF { _1 = a, _2 = b })
    , Mul = λ(a : t) → λ(b : t) → fix (E.MulF { _1 = a, _2 = b })
    }

-- Example usage.
let result
  = λ(t : Type)
  → λ(fix : ExprF t → t)
  → let E = Expr t fix in
    E.Add
      (E.Mul (E.Lit 3) (E.Lit 7))
      (E.Add (E.Lit 1) (E.Lit 2))

in result
```
We're able to interpret it from GHCi:

```haskell
> input @Expr auto "./Expr.dhall"
Add (Mul (Lit 3) (Lit 7)) (Add (Lit 1) (Lit 2))
```

Within any sort of non trivial configuration you'll sooner or later want to encode some recursive data structure, so an utility like that is a must.

So my question here is:
  * Should I make it into a pull request into the dhall-haskell repo?
  * Should I make it into a standalone library?
  * Or will you provide an alternative solution to this problem?

And the second part of this request is about adding functionality to dhall-json that would allow admitting values of the above type and convert them to recursive json representation. (possibly behind a flag/option)

**Comment by MaxOw** [comment]
Pinging @Gabriel439 to check if this issue was overlooked/folks are busy, or if there is simply no interest and I should close this?

**Comment by Gabriel439** [comment]
@MaxOw: Sorry for the delay.  I was just working my way through my inbox

Yeah, I would accept a pull request for this.  I can also help with the α-renaming issues during the pull request review.

One of the easiest tricks for dealing with variable names when matching against a syntax tree is to first `Dhall.Core.alphaNormalize` the tree and then look for specific indices.  You can see an example of this trick in the `dhall-json` code for translating the Dhall encoding of a recursive JSON structure to JSON.  It first `alphaNormalize`s the expression here:

https://github.com/dhall-lang/dhall-haskell/blob/42d80b057d560948b33d7194883b46eeaf002681/dhall-json/src/Dhall/JSONToDhall.hs#L377

... and then anywhere it needs to match a specific variable name it can guarantee that the variable will be of the form `_@n` for a fixed `n`, like it does here:

https://github.com/dhall-lang/dhall-haskell/blob/42d80b057d560948b33d7194883b46eeaf002681/dhall-json/src/Dhall/JSONToDhall.hs#L502-L513

In other words, it doesn't have to even look at the `Pi`-bound variable because it knows it will be `_` after α-normalization.

**Comment by MaxOw** [comment]
OK, I'm going to try to put it together into something more proper and make a PR.

**Comment by MaxOw** [comment]
Resolved via PR #1195. The second part of this request (about the same thing for json) I don't need right now but may tackle at some point in the future, if no one else will step up. Anyway, closing I'm closing this issue for now.
